[WIP] Tectonic-dependent dynamic friction weakening

Minerallo opened the pull request:
Hey everyone,

This PR adds a differential dynamic-friction mechanism whose high-strain-rate friction target depends on whether the tectonic regime is convergent, neutral/transform, or divergent. The regime is classified from the divergence of tangential velocity along the top boundary, smoothed, and projected through a configurable lithospheric depth. The implementation also exposes `tectonic_divergence_indicator` and `tectonic_regime` (`-1`, `0`, or `+1`) as named material-model outputs.

In global models, plastic strain weakening is often unreliable at low resolution due to poor strain localization, which leads to an underestimation of lithospheric weakening. To initiate plate-like behavior, models frequently compensate by reducing the frictional strength to unrealistically low values. As an alternative, dynamic friction weakening, where the friction angle decreases with increasing strain rate, enables a more physically realistic representation of lithospheric strength. However, the minimum strength of plate boundaries may vary depending on the tectonic setting (e.g., melt-induced weakening at mid-ocean ridges or fluid-assisted weakening in subduction zones). To address this, I have implemented a new friction weakening model that uses the trace of the strain rate tensor to distinguish between convergent and divergent regions. This allows the model to apply different minimum friction angles based on tectonic context.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.


For the test case;
the test run the new `tests.differential_dynamic_friction_surface_divergence` regression, and the existing `tests.dynamic_friction` regression. The new GNUplot test verifies convergent (`-1`, 2 degrees), neutral (`0`, 5 degrees), and divergent (`+1`, 1 degree) regions in one continuous incompressible velocity field.

## Results

### Prescribed-Stokes regression test

<img width="1944" height="1386" alt="strain_rate_weakening_regimes" src="https://github.com/user-attachments/assets/c4114b31-9073-4277-8a5f-786bef2b1e69" />

The prescribed incompressible velocity field produces convergent (`-1`, blue), neutral (`0`, white), and divergent (`+1`, red) regimes. The projected regimes select friction angles of 2, 5, and 1 degrees, respectively.

### Annulus application example

<img width="1152" height="841" alt="strain_rate_weakening_regimes_anulus" src="https://github.com/user-attachments/assets/542aa45c-a3c7-4f16-b797-02f35d3df0c7" />


also here is an annulus model that illustrates the projected surface tectonic regimes and the resulting spatial distribution of the current friction angle (this is not part of the test), In this example convergence happens on top of the subduction and the current friction angle is reduced at ~1degree.

Minerallo committed:
add differential strain rate weakening

Minerallo committed:
add Differential strain rate weakening case

Minerallo committed:
Corrected variable type

Minerallo committed:
Add surface-divergence-dependent dynamic friction

Minerallo committed:
run indent and param

Minerallo committed:
Reduce direct Boost usage

Minerallo committed:
Update differential friction test output

Minerallo commented:
Hi everyone, I think this is in good state to be checked maybe @tjhei since we discussed about it, if not @naliboff, @anne-glerum, @gassmoeller  ? thank you !

tjhei commented on tests/differential_dynamic_friction_surface_divergence.prm:
```diff
@@ -0,0 +1,83 @@
+# Enable if: QUICK_TEST
```
maybe remove this?

tjhei commented:
> `tectonic_divergence_indicator`

What values can this be? Is this an indicator or the surface divergence (floating point number)? The name should reflect that.

Can you please add a changelog entry?

Otherwise good from my point of view.

Minerallo committed:
Clarify surface divergence output

Minerallo commented:
> > `tectonic_divergence_indicator`
> 
> What values can this be? Is this an indicator or the surface divergence (floating point number)? The name should reflect that.
> 
> Can you please add a changelog entry?
> 
> Otherwise good from my point of view.

thanks @tjhei, I changed the name from tectonic_divergence_indicator for the surface_velocity_divergence, I added an entry to the changelog too, and here is an image for the first figure with the surface_velocity_divergence. 

<img width="1960" height="680" alt="image" src="https://github.com/user-attachments/assets/541ea69e-45e3-4caf-969e-cb835871816f" />

cheers
